Framework/src/InfrastructureGenerator.cxx

```diff
@@ -368,47 +370,34 @@ void InfrastructureGenerator::generateCheckRunners(framework::WorkflowSpec& work
         // Create a grouping key - sorted vector of stringified InputSpecs
         std::sort(inputNames.begin(), inputNames.end());
         // Group checks
-        checkRunnerMap[inputNames].push_back(check);
+        checksMap[inputNames].push_back(check);
       }
     }
   }
 
-  // For every Task output, find device to store the MO
-  for (auto& [label, inputSpec] : checkInputMap) {
+  // For every Task output, find a Check to store the MOs in the database.
+  // If none is found we create a sink device.
+  for (auto& [label, inputSpec] : tasksOutputMap) { // for each task output
     (void)inputSpec;
-    bool isStored = false;
-    // Look for single input
-    for (auto& [inputNames, checks] : checkRunnerMap) {
+    // Look for this task as input in the Checks' inputs, if we found it then we are done
+    for (auto& [inputNames, checks] : checksMap) { // for each set of inputs
       (void)checks;
       if (std::find(inputNames.begin(), inputNames.end(), label) != inputNames.end() && inputNames.size() == 1) {
         storeVectorMap[inputNames].push_back(label);
-        isStored = true;
         break;
       }
     }
 
-    if (!isStored) {
-      // If not assigned to store in previous step, find a candidate without input size limitation
-      for (auto& [inputNames, checks] : checkRunnerMap) {
-        (void)checks;
-        if (std::find(inputNames.begin(), inputNames.end(), label) != inputNames.end()) {
-          storeVectorMap[inputNames].push_back(label);
-          isStored = true;
-          break;
-        }
-      }
-    }
-
-    if (!isStored) {
-      // If there is no Check for a given input, create a candidate for a sink device
-      InputNames singleEntry{ label };
-      // Init empty Check vector to appear in the next step
-      checkRunnerMap[singleEntry];
-      storeVectorMap[singleEntry].push_back(label);
-    }
+    // If there is no Check for a given input, create a candidate for a sink device
+    InputNames singleEntry{ label };
+    // Init empty Check vector to appear in the next step
+    checksMap[singleEntry];
+    storeVectorMap[singleEntry].push_back(label);
```

Comment:
I don't understand why this can be put outside of the if statement.

Reply:
Now that I look at it with fresh eyes, I wonder the same. I will restore it.

Reply:
OK. Also that first loop was doubled before, but it seemed actually redundant. Still, I would give it a go with a few configurations just in case.

Reply:
I tried with a few different ones, with and without checks, and it seems to work. I am now considering adding tests to replace the old ones that were not testing anything anymore.

Reply:
Thanks. Actually, could you please make the refactoring of `InfrastructureGenerator::generateCheckRunners` a separate commit, in case we have to revert it?

Reply:
agreed
on it